### ISSUE TITLE
StackOverflow in MarkTypeForDynamicallyAccessedMembers

### ISSUE THREAD

**Issue by mdh1418** [mono/linker#1881]
Building `System.ComponentModel.Annotations/tests/System.ComponentModel.Annotations.Tests.csproj` for iOS with `/p:EnableAggressiveTrimming=true` leads to a Stack Overflow `error MSB6006: "dotnet" exited with code 134`.

### trailing section of error message
```
   at Mono.Linker.Steps.MarkStep.MarkEntireTypeInternal(Mono.Cecil.TypeDefinition, Boolean, Boolean, Mono.Linker.DependencyInfo ByRef, Mono.Cecil.IMemberDefinition, System.Collections.Generic.HashSet`1<Mono.Cecil.TypeDefinition>)
   at Mono.Linker.Steps.MarkStep.MarkEntireType(Mono.Cecil.TypeDefinition, Boolean, Boolean, Mono.Linker.DependencyInfo ByRef, Mono.Cecil.IMemberDefinition)
   at Mono.Linker.Dataflow.ReflectionMethodBodyScanner+<>c__DisplayClass26_1.<MarkTypeForDynamicallyAccessedMembers>b__0()
   at Mono.Linker.Dataflow.ReflectionMethodBodyScanner.MarkTypeForDynamicallyAccessedMembers(Mono.Linker.Dataflow.ReflectionPatternContext ByRef, Mono.Cecil.TypeDefinition, System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes)
   at Mono.Linker.Dataflow.ReflectionMethodBodyScanner.RequireDynamicallyAccessedMembers(Mono.Linker.Dataflow.ReflectionPatternContext ByRef, System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes, Mono.Linker.Dataflow.ValueNode, Mono.Cecil.IMetadataTokenProvider)
   at Mono.Linker.Dataflow.ReflectionMethodBodyScanner.ProcessAttributeDataflow(Mono.Cecil.IMemberDefinition, Mono.Cecil.MethodDefinition, System.Collections.Generic.IList`1<Mono.Cecil.CustomAttributeArgument>)
   at Mono.Linker.Steps.MarkStep.MarkCustomAttributeArguments(Mono.Cecil.IMemberDefinition, Mono.Cecil.CustomAttribute)
   at Mono.Linker.Steps.MarkStep.MarkCustomAttribute(Mono.Cecil.CustomAttribute, Mono.Linker.DependencyInfo ByRef, Mono.Cecil.IMemberDefinition)
   at Mono.Linker.Steps.MarkStep.MarkCustomAttributes(Mono.Cecil.ICustomAttributeProvider, Mono.Linker.DependencyInfo ByRef, Mono.Cecil.IMemberDefinition)
   at Mono.Linker.Steps.MarkStep.MarkEntireTypeInternal(Mono.Cecil.TypeDefinition, Boolean, Boolean, Mono.Linker.DependencyInfo ByRef, Mono.Cecil.IMemberDefinition, System.Collections.Generic.HashSet`1<Mono.Cecil.TypeDefinition>)
   at Mono.Linker.Steps.MarkStep.MarkEntireType(Mono.Cecil.TypeDefinition, Boolean, Boolean, Mono.Linker.DependencyInfo ByRef, Mono.Cecil.IMemberDefinition)
   at Mono.Linker.Dataflow.ReflectionMethodBodyScanner+<>c__DisplayClass26_1.<MarkTypeForDynamicallyAccessedMembers>b__0()
   at Mono.Linker.Dataflow.ReflectionMethodBodyScanner.MarkTypeForDynamicallyAccessedMembers(Mono.Linker.Dataflow.ReflectionPatternContext ByRef, Mono.Cecil.TypeDefinition, System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes)
   at Mono.Linker.Dataflow.ReflectionMethodBodyScanner.RequireDynamicallyAccessedMembers(Mono.Linker.Dataflow.ReflectionPatternContext ByRef, System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes, Mono.Linker.Dataflow.ValueNode, Mono.Cecil.IMetadataTokenProvider)
   at Mono.Linker.Dataflow.ReflectionMethodBodyScanner.ProcessAttributeDataflow(Mono.Cecil.IMemberDefinition, Mono.Cecil.MethodDefinition, System.Collections.Generic.IList`1<Mono.Cecil.CustomAttributeArgument>)
   at Mono.Linker.Steps.MarkStep.MarkCustomAttributeArguments(Mono.Cecil.IMemberDefinition, Mono.Cecil.CustomAttribute)
   at Mono.Linker.Steps.MarkStep.MarkCustomAttribute(Mono.Cecil.CustomAttribute, Mono.Linker.DependencyInfo ByRef, Mono.Cecil.IMemberDefinition)
   at Mono.Linker.Steps.MarkStep.MarkCustomAttributes(Mono.Cecil.ICustomAttributeProvider, Mono.Linker.DependencyInfo ByRef, Mono.Cecil.IMemberDefinition)
   at Mono.Linker.Steps.MarkStep.MarkType(Mono.Cecil.TypeReference, Mono.Linker.DependencyInfo, Mono.Cecil.IMemberDefinition)
   at Mono.Linker.Steps.MarkStep.MarkCustomAttributeArgument(Mono.Cecil.CustomAttributeArgument, Mono.Cecil.ICustomAttribute, Mono.Cecil.IMemberDefinition)
   at Mono.Linker.Steps.MarkStep.MarkCustomAttributeArguments(Mono.Cecil.IMemberDefinition, Mono.Cecil.CustomAttribute)
   at Mono.Linker.Steps.MarkStep.MarkCustomAttribute(Mono.Cecil.CustomAttribute, Mono.Linker.DependencyInfo ByRef, Mono.Cecil.IMemberDefinition)
   at Mono.Linker.Steps.MarkStep.MarkCustomAttributes(Mono.Cecil.ICustomAttributeProvider, Mono.Linker.DependencyInfo ByRef, Mono.Cecil.IMemberDefinition)
   at Mono.Linker.Steps.MarkStep.MarkEntireTypeInternal(Mono.Cecil.TypeDefinition, Boolean, Boolean, Mono.Linker.DependencyInfo ByRef, Mono.Cecil.IMemberDefinition, System.Collections.Generic.HashSet`1<Mono.Cecil.TypeDefinition>)
   at Mono.Linker.Steps.MarkStep.MarkEntireTypeInternal(Mono.Cecil.TypeDefinition, Boolean, Boolean, Mono.Linker.DependencyInfo ByRef, Mono.Cecil.IMemberDefinition, System.Collections.Generic.HashSet`1<Mono.Cecil.TypeDefinition>)
   at Mono.Linker.Steps.MarkStep.MarkEntireAssembly(Mono.Cecil.AssemblyDefinition)
   at Mono.Linker.Steps.MarkStep.MarkAssembly(Mono.Cecil.AssemblyDefinition, Mono.Linker.DependencyInfo)
   at Mono.Linker.Steps.MarkStep.MarkModule(Mono.Cecil.ModuleDefinition, Mono.Linker.DependencyInfo)
   at Mono.Linker.Steps.MarkStep.ProcessMarkedPending()
   at Mono.Linker.Steps.MarkStep.Initialize()
   at Mono.Linker.Steps.MarkStep.Process(Mono.Linker.LinkContext)
   at Mono.Linker.Pipeline.ProcessStep(Mono.Linker.LinkContext, Mono.Linker.Steps.IStep)
   at Mono.Linker.Pipeline.Process(Mono.Linker.LinkContext)
   at Mono.Linker.Driver.Run(Mono.Linker.ILogger)
   at Mono.Linker.Driver.Main(System.String[])
./linker_command.sh: line 1: 59610 Abort trap: 6
```

Part of https://github.com/dotnet/runtime/pull/49390
Error log, search for `Stack overflow` https://dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_apis/build/builds/1029931/logs/86

Zip file repro steps
Unzip the following and modify the first part of `linker_command.sh` `/Users/mdhwang/runtime_wasm_clean/.dotnet/dotnet "/Users/mdhwang/.nuget/packages/microsoft.net.illink.tasks/6.0.100-preview.2.21126.1/tools/net5.0/illink.dll"` to conform to local dotnet and illink.
[iOS_EnableAggressiveTrimming_linker_bug.zip](https://github.com/mono/linker/files/6117702/iOS_EnableAggressiveTrimming_linker_bug.zip)

/cc:@marek-safar

**Comment by marek-safar** [comment]
@vitek-karas @MichalStrehovsky there seems to be recursion in MarkTypeForDynamicallyAccessedMembers

**Comment by vitek-karas** [comment]
A "simple" repro looks like this:

```C#
		class AllOnSelf
		{
			public static void Test ()
			{
				var t = typeof (KeepAllOnSelf);
			}

			class KeepsAllAttribute : Attribute
			{
				public KeepsAllAttribute (
					[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.All)]
					Type type)
				{
				}
			}

			[KeepsAll(typeof(KeepAllOnSelf))]
			class KeepAllOnSelf
			{
				public void Method () { }
				public int Field;
			}
		}
```

Attribute which applies `All` on its parameter. Class which adds this attribute referencing itself in it. We will call `MarkEntireType` on the attribute parameter, which in turn will try to mark all the attributes on the type, which in turn will apply `All` to the attribute parameter, which calls `MarkEntireType`.

I can't think of another way to solve this other than keeping a "Global" hashset of types marked as entire. We already maintain a hashset like that during the `MarkEntireType` to avoid recursion caused by interfaces and base types. Passing it through all of the dependencies is not practical (there would be probably something like 50 methods to pass this through). The overall number of types in the hashset should be small enough.

**Comment by MichalStrehovsky** [comment]
> I can't think of another way to solve this other than keeping a "Global" hashset of types marked as entire.

Yep, that look like the par for the course in MarkStep.

The .NET Native dependency analyzer had a better design around this that basically had one HashSet and one Queue:

```csharp
void Process()
{
    AddRoots();

    while (Dequeue(out x))
        Scan(x);
}

void Mark(Type x, Kind kind)
{
    if (AlreadyProcessed(x, kind))
        return;

    Enqueue(x, kind);
}

void Scan(x)
{
    if (AlreadyProcessed(x))
        return;

    Mark(x.BaseType);

    // ...

    AddToProcessed(x);
}
```

If someone feels like doing another refactoring of MarkStep this (one hashset and one queue) would be the direction to go to - it avoids all the weird local hashsets, a separate "marked" and "processed" state, etc. These things greatly confuse me every time I look at them. It would be a reduction in complexity, bugfarming potential, and memory use.

**Comment by vitek-karas** [comment]
Fixed in #1886
